### PR TITLE
AnnotationSuppressor now resolves Full Qualified Annotation Names without type solving

### LINKED ISSUE #4355: LongParameterList no longer recognizes ignored annotations in 1.19.0
## Expected Behavior

Prior to 1.19.0, this configuration worked to prevent `LongParameterList` warnings on constructors with a few "known" injection annotations from dagger.

```yml
LongParameterList:
  active: true
  functionThreshold: 7
  constructorThreshold: 7
  ignoreDefaultParameters: false
  ignoreDataClasses: true
  ignoreAnnotated: ['javax.inject.Inject', 'dagger.Provides', 'dagger.assisted.AssistedInject']
```

## Observed Behavior

In 1.19.0, these seem to be ignored now

## Steps to Reproduce

Here are some snippets that repro for us

```kotlin
class ActivityFeedPresenter @Inject constructor(
  private val activityApi: ActivityApi,
  private val userRepository: UserRepository,
  private val botsDataProvider: BotsDataProvider,
  private val mpdmDisplayNameHelper: MpdmDisplayNameHelper,
  private val prefsManager: PrefsManager,
  private val conversationRepo: ConversationRepository,
  private val lockedThreadTransformer: ActivityFeedLockedThreadTransformer,
  private val mentionItemFactoryLazy: Lazy<MentionItemFactory>,
) : ActivityFeedContract.Presenter, LoadingStateProvider {

class AppDialogFragment @AssistedInject constructor(
  private val appDialogsRepository: AppDialogsRepository,
  private val presenter: AppDialogPresenter?,
  private val keyboardHelper: KeyboardHelper,
  private val snackbarHelper: SnackbarHelper,
  private val appBuildConfig: AppBuildConfig,
  private val appDialogMenuBinder: AppDialogMenuBinder,
  private val appDialogMenuViewFactory: AppDialogMenuView.Factory,
  private val appDialogSelectViewFactory: AppDialogSelectView.Factory,
  private val appDialogTextViewFactory: AppDialogTextView.Factory,
  private val platformLogger: PlatformLogger,
  @ForScope(UserScope::class) private val fragmentNavRegistrar: FragmentNavRegistrar
) : ViewBindingFragment(),
  AppDialogContract.View,
  MenuSelectionListener,
  BackPressedListener {
  
class AppActionDelegateImpl @Inject constructor(
  private val appActionsRepository: Lazy<AppActionsRepository>,
  private val attachmentActionHelperLazy: Lazy<AttachmentActionHelper>,
  private val attachmentRepositoryLazy: Lazy<AttachmentRepository>,
  private val blockKitActionDelegateLazy: Lazy<BlockKitActionDelegate>,
  private val blockKitDialogHelperLazy: Lazy<BlockKitDialogHelper>,
  private val customTabHelperLazy: Lazy<CustomTabHelper>,
  private val formattedTextBinderLazy: Lazy<FormattedTextBinder>,
  private val messageActionsHelperLazy: Lazy<MessageActionsHelper>,
  private val commandRepository: Lazy<CommandRepository>,
  private val platformEventHandler: Lazy<PlatformEventHandler>,
  private val toasterLazy: Lazy<Toaster>,
  private val typefaceSubstitutionHelperLazy: Lazy<TypefaceSubstitutionHelper>,
  private val networkInfoManagerLazy: Lazy<NetworkInfoManager>,
  private val platformAppsManagerLazy: Lazy<PlatformAppsManager>,
  private val messageDaoLazy: Lazy<MessageDao>,
  private val messageEventBroadcasterLazy: Lazy<MessageEventBroadcaster>,
  private val slackActionDelegateLazy: Lazy<SlackActionDelegate>,
  private val snackbarDelegate: SnackbarDelegate,
  private val blockKitStateProvider: BlockKitStateProvider,
  private val triggerResolver: TriggerResolver,
  @ForLegacyFeature(MOBILE_INPUT_BLOCKS_IN_MESSAGES) private val isMobileInputBlocksInMessagesEnabled: Boolean
) : AppActionDelegate, SnackbarDelegate by snackbarDelegate {
```

## Context

It is a regression

## Your Environment

* Version of detekt used: 1.19.0
* Version of Gradle used (if applicable): N/A
* Gradle scan link (add `--scan` option when running the gradle task): private repo
* Operating System and version: macOS 11
* Link to your project (if it's a public repository): N/A

### LINKED ISSUE #4356: ignoreAnnotated not respected on UnnecessaryAbstractClass
## Expected Behavior

After migrating to `ignoreAnnotated` per deprecation warning messages, it's no longer respecting the annotations we specify

Detekt 1.18.0
```yml
UnnecessaryAbstractClass:
  active: true
  excludeAnnotatedClasses: ['dagger.Module']
```

Warns

```
Property 'style>UnnecessaryAbstractClass>excludeAnnotatedClasses' is deprecated. Use `ignoreAnnotated` instead.
```

Detekt 1.19.0
```yml
UnnecessaryAbstractClass:
  active: true
  ignoreAnnotated: ['dagger.Module']
```

## Observed Behavior

It incorrectly warns about classes annotated with `dagger.Module`

## Steps to Reproduce

Here's an example module

```kotlin
@Module
abstract class CallsBaseModule {

  @Binds
  abstract fun CallPresenterImpl.provideCallPresenter(): CallContract.CallPresenter

  @Binds
  abstract fun CallServiceProviderImpl.bindCallServiceProvider(): CallServiceProvider
}
```

## Context

It is a regression

## Your Environment

* Version of detekt used: 1.19.0
* Version of Gradle used (if applicable): N/A
* Gradle scan link (add `--scan` option when running the gradle task): N/A
* Operating System and version: macOS 11
* Link to your project (if it's a public repository): N/A

### PR DESCRIPTION
**Pull request by BraisGabin**
This was way more difficult than I expected. Please, review this PR commit by commit.

The more complex type of annotations are like this `@com.dagger.Component.Factory`. We have a part that is package and another part that are class names. But it's not clear when one ends and one the other ends. This is importan because when we find for matches for this type of annotations we should allow:

- `@com.dagger.Component.Factory`
- `@Component.Factory` if `import com.dagger.Component is present`
- `@Factory` if `import com.dagger.Component.Factory is present`

But we should not allow `@dagger.Component.Factory`, because that's another annotation different. It can't be `@com.dagger.Component.Factory`. All this cases is what makes this code complex and that force me to have all those tests.

fixes #4355
fixes #4356

### CODE REVIEW

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt [hunk 1]
```diff
@@ -11,48 +15,82 @@ import org.jetbrains.kotlin.psi.KtFile
...
-        .split(".")
-        .dropWhile { it.first().isLowerCase() }
-        .reversed()
-        .scan("") { acc, name ->
-            if (acc.isEmpty()) name else "$name.$acc"
-        }
-        .drop(1) + fqName
+private fun FqName.getPackage(): String {
+    // This is a shortcut. We should make this properly
+    return this.toString().getPackage()
```
Comment (BraisGabin) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt:73]: This is wrong. We should be able to know what part of the `FqName` is the `package` name and witch part is the `class` name. But to be honest, I don't know how to extract that information so I use the same heuristic that we use when we don't have type-solving.

I created and skipped a test that should be enabled once this is fixed.

Comment (cortinico) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt:73]: Can you extend a bit the comment you left in the code with some of the context you provided here.

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt [hunk 2]
```diff
@@ -42,33 +45,52 @@ class AnnotationExcluder(
...
-         @dagger.Component.Factory
-         For that reason we use a heuristic here: If the first character is lower case we assume it's a package name
-         */
-        val possibleNames = if (!annotationText.first().isLowerCase()) {
-            fullQualifiedNameGuesser.getFullQualifiedName(annotationText)
+    private fun isExcluded(annotation: KtTypeReference, context: BindingContext): Boolean {
+        val fqName = annotation.fqNameOrNull(context)
+        return if (fqName == null) {
+            fullQualifiedNameGuesser.getFullQualifiedName(annotation.text.toString())
+                .map { it.getPackage() to it }
```
Comment (cortinico) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt:56]: I think here we should either:
- Add an explicitly `if (context == BindingContext.EMPTY)` check
- Left a comment that this is the non-type resolution scenario.

As it's not immediately clear that this information is carried by the `fqNameOrNull` function

**Comment by cortinico** [comment]
> This was way more difficult than I expected. Please, review this PR commit by commit.

I went through them and I lack the relevant context as I haven't really touched this part of the codebase in a while. The code looks good and well tested. Generally, I think we're good to go with the heuristic you suggested.

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt (hunk 2 above)
Comment (BraisGabin) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt]: Done

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt (hunk 1 above)
Comment (BraisGabin) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt:77]: done

File: detekt-psi-utils/src/test/kotlin/io/github/detekt/psi/internal/FullQualifiedNameGuesserSpec.kt [hunk 3]
```diff
@@ -61,6 +61,12 @@ class FullQualifiedNameGuesserSpec {
             assertThat(sut.getFullQualifiedName("Static.Factory"))
                 .containsExactlyInAnyOrder("kotlin.jvm.JvmStatic.Factory")
         }
+
+        @Test
+        fun `when don't use the import`() {
```
Comment (schalkms) [inline detekt-psi-utils/src/test/kotlin/io/github/detekt/psi/internal/FullQualifiedNameGuesserSpec.kt:66]:
```suggestion
        fun `when not using the import`() {
```

**Review by schalkms — APPROVED** [review]

File: detekt-psi-utils/src/main/kotlin/io/github/detekt/psi/internal/FullQualifiedNameGuesser.kt [hunk 4]
```diff
@@ -39,13 +39,16 @@ class FullQualifiedNameGuesser internal constructor(
                     if (packageName != null) {
                         add("$packageName.$name")
                     }
+                    if (name.first().isLowerCase()) {
```
Comment (schalkms) [inline detekt-psi-utils/src/main/kotlin/io/github/detekt/psi/internal/FullQualifiedNameGuesser.kt:42]: What's the reason for this check?

**Comment by schalkms** [comment]
The drawback of having this mixed implementation for enabled and disabled type resolution is the complexity of the code. However, for now, I don't know a better solution.

**Comment by 3flex** [comment]
> The drawback of having this mixed implementation for enabled and disabled type resolution is the complexity of the code

Agree. This approach will lead to significant maintenance if for every rule or feature that's improved by using type resolution we have to also support users who choose not to use type resolution.

Is the real issue that type resolution is too difficult to use correctly? What can we do here to avoid introducing this complexity to our code?

**Comment by BraisGabin** [comment]
> Is the real issue that type resolution is too difficult to use correctly? What can we do here to avoid introducing this complexity to our code?

As far as I know the main issue is the performance. We should rework our gradle plugin to make it type solve first.

**Comment by cortinico** [comment]
> As far as I know the main issue is the performance

This ☝️ . We have users running detekt on pre-commit hooks. Enabling TR for them is a no-go as they'll end up impacting the development loop for a lot of engineers.

I think we should keep the current feature set for non-TR enabled project and add new features that are working with TR only.

The problem in this PR/Linked issues was that, by implementing a more advanced annotation suppressor, we actually broke the experience for users on non-TR project.

**Comment by cortinico** [comment]
Anything missing from this PR? I'd like to kickoff a RC1 as soon as this is merged 👍

**Comment by BraisGabin** [comment]
This is ready to merge for me. But, is this a notable change or just a bug fix?

File: detekt-api/src/test/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluderSpec.kt [hunk 5]
```diff
@@ -1,124 +1,170 @@
...
+    }
+
+    @Nested
+    inner class `difference between type solving and no type solving` {
+
         @Nested
-        inner class `special cases` {
-            private val annotation = psiFactory.createAnnotationEntry("@Component")
-            private val sinceKotlinAnnotation = psiFactory.createAnnotationEntry("@SinceKotlin")
+        inner class `Don't mix annotations with the same name` {
```
Comment (chao2zhang) [inline detekt-api/src/test/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluderSpec.kt:96]: I guess if people use star imports, that would also be a case where no type resolving fails.

**Review by chao2zhang — APPROVED** [review]
lgtm! Thank you for adding those test cases and that really helped me to understand

File: detekt-api/src/test/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluderSpec.kt (hunk 5 above)
Comment (BraisGabin) [reply detekt-api/src/test/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluderSpec.kt:96]: You are right! I'm not handling the star imports in `FullQualifiedNameGuesser`. I'll try to fix this asap to release RC1 asap.

So much work to keep mixed behaviours >_<.

File: detekt-psi-utils/src/main/kotlin/io/github/detekt/psi/internal/FullQualifiedNameGuesser.kt (hunk 4 above)
Comment (BraisGabin) [reply detekt-psi-utils/src/main/kotlin/io/github/detekt/psi/internal/FullQualifiedNameGuesser.kt:42]: Sorry @schalkms, I didn't see this comment before.

Yes the reason was that heuristic... but now that I think about that I'm not sure if this class needs to know about this. We could assume that any string that we can't match with an import could be a FullQualifiedName. But I don't know if that's better or worse... That's the issue with this mixed behaviour. The "general use cases" are easy to manage but all this edge cases are really difficult to manage because we can't know for sure.

File: detekt-api/src/test/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluderSpec.kt (hunk 5 above)
Comment (BraisGabin) [reply detekt-api/src/test/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluderSpec.kt:96]: Done

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt [hunk 6]
```diff
@@ -11,48 +15,86 @@ import org.jetbrains.kotlin.psi.KtFile
...
-        val possibleNames = if (!annotationText.first().isLowerCase()) {
-            fullQualifiedNameGuesser.getFullQualifiedName(annotationText)
+    fun shouldExclude(annotations: List<KtAnnotationEntry>): Boolean {
+        return annotations.any { annotation -> annotation.typeReference?.let { isExcluded(it, context) } ?: false }
+    }
+
+    private fun isExcluded(annotation: KtTypeReference, context: BindingContext): Boolean {
+        val fqName = if (context == BindingContext.EMPTY) null else annotation.fqNameOrNull(context)
+        return if (fqName == null) {
+            fullQualifiedNameGuesser.getFullQualifiedName(annotation.text.toString())
```
[the remark this answers is not in the document]
Comment (BraisGabin) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt:55]: I know. For this reason I wrote all those tests. To have them as kind of documentation... I don't know if adding some comments in this code explaining why they are needed would help. I'm not a fan of code with comments becuase those comments get outdated very easy but maybe here it have sense.

[the remark this answers is not in the document]
Comment (BraisGabin) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt:55]: > Sounds like the `if-else` here to could be assigned to `possibleNames` like before

Done

**Comment by cortinico** [comment]
> This is ready to merge for me. But, is this a notable change or just a bug fix?

It's a notable change that we fixed a regression in the latest stable (imho this would have been point-release worthy, we skipped that so let's at least mention this on top of the release notes).

**Review by chao2zhang — APPROVED** [review]
